streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogTopics.java

```diff
@@ -0,0 +1,132 @@
+    }
+
+    public void setup() {
+        // add tasks to state change log topic subscribers
+        final Map<String, InternalTopicConfig> changelogTopicMetadata = new HashMap<>();
+        for (final Map.Entry<Integer, TopicsInfo> entry : topicGroups.entrySet()) {
+            final int topicGroupId = entry.getKey();
+            final TopicsInfo topicsInfo = entry.getValue();
+
+            final Set<TaskId> topicGroupTasks = tasksForTopicGroup.get(topicGroupId);
+            if (topicGroupTasks == null) {
+                log.debug("No tasks found for topic group {}", topicGroupId);
+                continue;
+            } else if (topicsInfo.stateChangelogTopics.isEmpty()) {
+                continue;
+            }
+
+            for (final TaskId task : topicGroupTasks) {
+                final Set<TopicPartition> changelogTopicPartitions = topicsInfo.stateChangelogTopics
+                    .keySet()
+                    .stream()
+                    .map(topic -> new TopicPartition(topic, task.partition))
+                    .collect(Collectors.toSet());
+                changelogPartitionsForTask.put(task, changelogTopicPartitions);
+            }
+
+            for (final InternalTopicConfig topicConfig : topicsInfo.nonSourceChangelogTopics()) {
+                // the expected number of partitions is the max value of TaskId.partition + 1
+                int numPartitions = UNKNOWN;
+                for (final TaskId task : topicGroupTasks) {
+                    if (numPartitions < task.partition + 1) {
+                        numPartitions = task.partition + 1;
+                    }
+                }
+                topicConfig.setNumberOfPartitions(numPartitions);
+                changelogTopicMetadata.put(topicConfig.name(), topicConfig);
+            }
+            sourceTopicBasedChangelogTopics.addAll(topicsInfo.sourceTopicChangelogs());
+        }
+
+        final Set<String> newlyCreatedChangelogTopics = internalTopicManager.makeReady(changelogTopicMetadata);
+        log.debug("Created state changelog topics {} from the parsed topology.", changelogTopicMetadata.values());
+
+        for (final Map.Entry<TaskId, Set<TopicPartition>> entry : changelogPartitionsForTask.entrySet()) {
+            final TaskId taskId = entry.getKey();
+            final Set<TopicPartition> topicPartitions = entry.getValue();
+            for (final TopicPartition topicPartition : topicPartitions) {
+                if (!newlyCreatedChangelogTopics.contains(topicPartition.topic())) {
+                    preExistingChangelogPartitionsForTask.computeIfAbsent(taskId, task -> new HashSet<>()).add(topicPartition);
+                    if (!sourceTopicBasedChangelogTopics.contains(topicPartition.topic())) {
+                        preExistingNonSourceTopicBasedChangelogPartitions.add(topicPartition);
+                    } else {
+                        sourceTopicBasedChangelogTopicPartitions.add(topicPartition);
```

Comment:
Also here, if a changelog is source based it has to be pre-existing, because the `InternalTopicManager` will not setup source topics. However, making this clearer in the name does not harm.